Zend/zend_execute.c

```diff
@@ -1874,13 +1888,22 @@ static zend_always_inline void zend_fetch_dimension_address_read(zval *result, z
 			}
 		}
 	} else {
-		if (type != BP_VAR_IS && UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
-			zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+		if (type != BP_VAR_IS) {
+			if (UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
+				zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+			} else if (!zend_fetch_dimension_address_read_valid_container(container, support_strings)) {
+				zend_error(E_WARNING, "Variable of type %s does not accept array offsets", zend_get_type_by_const(Z_TYPE_P(container)));
+			}
 		}
 		if (/*dim_type == IS_CV &&*/ UNEXPECTED(Z_TYPE_P(dim) == IS_UNDEF)) {
 			zval_undefined_cv(EG(current_execute_data)->opline->op2.var, EG(current_execute_data));
 		}
-		ZVAL_NULL(result);
+
+		if (UNEXPECTED((EG(current_execute_data)->opline+1)->opcode == ZEND_FETCH_DIM_R && (EG(current_execute_data)->opline+1)->op1_type == IS_VAR)) {
```

Comment:
I suspect this will not work correctly for BP_FUNC_ARG fetches. E.g. try something like this

```php
$fn = 'var_dump';
$array = null;
$fn($array[0][0][0]);
```